proposals/0000-ghc-module-naming.rst

```diff
@@ -0,0 +1,145 @@
+
+
+Proposal 4
+------------
+
+* The public API of package ``ghc`` (GHC as a library) should have modules of form ``GhcAPI.*``.
+
+All of the modules in package ``ghc`` currently start with ``GHC.*`` which correctly signals that they are part of GHC's internals.
+As part of the GHC API redesign (a HF project in its own right, currently stalled) it would be very helpfult
+to modules with stable APIs, and a new prefix, such as ``GhcAPI.*``.
+
+
+Timescale
+==========
+The first release of GHC with `ghc-experimental` and `ghc-internals` will be GHC 9.10, which expect to
+release in early 2024.  It would be good to establish naming conventions for modules well before this date.
+
+Example lifecycle
+===================
+
+By way of example, consider the ``HasField`` class, which supports overloaded record fields.
+It is currently defined in ``base:GHC.Records``, which is an odd module to have to import.
+Moreover there is
+more than one GHC proposal that suggest changes to its design (e.g. see `GHC Proposal 158 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0158-record-set-field.rst>`_); it is not nearly as stable as most of ``base``
+
+If ``ghc-experimental`` had existed we would have put it in ``ghc-experimental:Experimental.Records``.
+That would have made it clear that the design of overloaded records still evolving.
+Once the design becomes settled and stable, it could move to ``base``, perhaps in a module like ``Data.Records``.
+
+Other similar examples include
+
+* The tuple proposal of `GHC Proposal 475 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst>`_
+* The `DataToTag CLC proposal <https://github.com/haskell/core-libraries-committee/issues/104>`_ would have been easier to expose through ``ghc-experimental`` in the first instance.
+
+Alternatives
+==============
+* We could dispute Proposal 1: one could imagine deliberately naming modules in ``ghc-experimental`` with the
+  same module name as their eventual expected (by someone) home in ``base``.  The goal would be to reduce impact if and when
+  the module moves from ``ghc-experimental`` to ``base``. For example, we might add ``Data.Tuple`` to ``ghc-experimental`` containing the new type constructors ``Tuple2``, ``Tuple3`` etc that are proposed in `GHC Proposal 475 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst>`_.   However:
+
+  * In the meantime there are two modules both called ``Data.Tuple``.  This is bad.  Which one does ``import Data.Tuple`` import?  (Look at the Cabal file, perhaps?)  How can I import both?  (Package-qualified imports perhaps.) So it will really only help in the case of a brand-new module, not already in ``base``.
+  * It loses the explicit cue, in the source code, given by ``import Experimental.Data.Tuple``.
+
+* We could use ``GHC.*`` for modules in ``ghc-experimental``, and maybe ``GHC.Internals.*`` for module in ``ghc-internals``.  But
```

Comment:
If we adopt the proposal of [having a module warning for the internal modules](https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst#discourging-the-direct-use-of-ghc-internals), then perhaps we don't need a `ghc-internals`-specific module prefix and we could just use `GHC`. Not sure if that's a good idea.

Reply:
That is true, but the "broader runtime" (which this is) vs the compiler itself are very different things. I think `GHC.Internals` isn't even different enough (say we wanted to use `GHC.Internals` for extra unstable parts of GHC itself?) but at least it is *something*.